db/connection/connection.go

```diff
@@ -117,6 +119,90 @@ type DB interface {
 	BulkInsert(ctx context.Context, tableName string, columns []string, values [][]interface{}) (execError error)
 }
 
+var _ DB = (*FlexibleTransaction)(nil)
+
+// FlexibleTransaction allows for a DB transaction to be passed through functions and avoid multiple commit/rollbacks
+// it also takes care of some of the most repeated checks at the time of commit/rollback and tx checking.
+type FlexibleTransaction struct {
+	DB
+	rolled               bool
+	concurrencySafeguard sync.Mutex
+}
+
+func (f *FlexibleTransaction) Cleanup(ctx context.Context) (bool, bool, error) {
+	f.concurrencySafeguard.Lock()
+	defer f.concurrencySafeguard.Unlock()
+	if f.DB == nil {
+		return false, false, nil
+	}
+	if f.rolled {
+		if err := f.DB.RollbackTransaction(ctx); err != nil {
+			return false, false, fmt.Errorf("rolling back transaction: %w", err)
+		}
+		return false, true, nil
+	}
+
+	if err := f.DB.CommitTransaction(ctx); err != nil {
+		return false, false, fmt.Errorf("committing transaction: %w", err)
+	}
+	return true, false, nil
+}
+
+// TXFinishFunc represents an all-encompassing function that either rolls or commits a tx based on the outcome.
+type TXFinishFunc func(ctx context.Context) (committed, rolled bool, err error)
+
+// BeginTransaction will wrap the passed DB into a transaction handler that supports it being used with less care
+// and prevents having to check if we are already in a tx and failures due to eager committers.
+func BeginTransaction(ctx context.Context, conn DB) (DB, TXFinishFunc, error) {
+	// this can happen so let's work around it
+	ft, isFT := conn.(*FlexibleTransaction)
+	if isFT {
+		return ft, func(ctx2 context.Context) (bool, bool, error) {
+			return false, false, nil
+		}, nil
+	}
+
+	// the underlying conn is a tx, let's be careful not to commit/rollback it
+	if conn.IsTransaction() {
+		return &FlexibleTransaction{
+				DB: conn,
+			},
+			func(ctx2 context.Context) (bool, bool, error) {
+				return false, false, nil
+			},
+			nil
+
+	}
+
+	tx, err := conn.BeginTransaction(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("beginning transaction: %w", err)
+	}
+
+	f := &FlexibleTransaction{
+		DB: tx,
+	}
+	return f, f.Cleanup, nil
+}
+
+// BeginTransaction implements DB for FlexibleTransaction
+func (f *FlexibleTransaction) BeginTransaction(ctx context.Context) (DB, error) {
+	return f, nil
+}
+
+// CommitTransaction implements DB for FlexibleTransaction
+func (f *FlexibleTransaction) CommitTransaction(ctx context.Context) error {
+	return nil
```

Comment:
I don't have experience designing APIs in Go but it seems weird to me, it can be really surprising that `db.Commit` (it's used polymorphically so it's not always apparent if you're working with regular DB or `FlexibleTransaction`) actually does nothing.

I think conceptually `DB` and say `Transaction` are different things, even if they share a lot of common operations. IMO having `type DB interface { Query, EQuery, ... BeginTransaction }` (no `Commit` and `Rollback`) and `type Transaction interface { Query, EQuery,... MarkToRollback` } would be more accurate.

Then code using `Transaction` wouldn't be able to `Commit` or `Rollback`, but it's also the case in this PR's code, with a difference that this PR's code would compile, creating an illusion that `FlexibleTransaction.Commit` actually does something. It would be able to `MarkToRollback`, also aligned with the PR's code, with a difference that it's more explicit that it only marks the transaction to be rolled back.

In other words, I am arguing that `FlexibleTransaction` introduces its own API: one that revolves around `TXFinishFunc` that has to be called. Nothing wrong with that but IMO keeping pretense that it also implements `DB` can be misleading.

Once again, just my 2 cents, I haven't even contributed to gaum so you're definitely in a better position to make a call. Especially, I don't know how feasible are big, possible incompatible changes in gaum, as my proposition would probably involve bigger change in gaum interfaces